hifiberrydsp/lg/soundsync.py

```diff
@@ -29,98 +30,110 @@
-        self.finished = False
-        self.pollinterval = 0.3
         self.spi = SpiHandler
-        self.dspdata = None
-        self.dspvol = None
-        self.lgsoundsyncdetected = False
+
+        self.finished = False
+        self.detected = False
+
+        self.volume_register = None
+        self.spdif_active_register = None
 
         Thread.__init__(self)
 
-    def set_registers(self, volume_register, spdif_register):
-        logging.info("LG soundsync: using volume at %s and SPDIF active at %s",
-                      volume_register, spdif_register)
+    def set_registers(self, volume_register, spdif_active_register):
+        logger.info("Using volume register at %s and SPDIF active register at %s",
+                    volume_register, spdif_active_register)
         self.volume_register = volume_register
-        self.spdif_register = spdif_register
+        self.spdif_active_register = spdif_active_register
 
     def update_volume(self):
-
-        if self.volume_register is None or self.spdif_register is None:
+        if self.volume_register is None or self.spdif_active_register is None:
             return False
 
-        # Read SPDIF status registers
-        data = self.spi.read(0xf617, 6)
-        if len(data) != 6:
-            logging.error("internal error: could not read 6 bytes from SPI")
+        if not self.is_spdif_active():
             return False
 
-        _b1, vol, volid, _b2 = struct.unpack(">BHHB", data)
+        volume = self.try_read_volume()
 
-        if volid != 0x048a:
+        if volume is None:
             return False
 
-        if vol < 0x100f or vol > 0x164f:
-            return False
+        self.write_volume(volume)
 
-        # Read SPDIF enable register
-        data = self.spi.read(self.spdif_register, 4)
-        [spdif_active] = struct.unpack(">l", data)
-        if spdif_active == 0:
-            return False
-
-        volpercent = (vol - 0x100f) / 16
-        if volpercent < 0 or volpercent > 100:
-            logging.error("internal error, got volume = %s, "
-                          "but should be in 0-100 range", volpercent)
-        # convert percent to multiplier
-        volume = percent2amplification(volpercent)
+        return True
 
-        # write multiplier to DSP
-        dspdata = datatools.int_data(self.dsp.decimal_repr(volume),
-                                     self.volume_register_length)
+    def is_spdif_active(self):
+        data = self.spi.read(self.spdif_active_register, 4)
+        [spdif_active] = struct.unpack(">l", data)
+        return spdif_active != 0
+
+    def try_read_volume(self):
+        spdif_status_register = 0xf617
+        return self.parse_volume_from_status(self.spi.read(spdif_status_register, 6))
+
+    # Volume          ~~~~~~~~~~~~~~~~~~~~           This is what the SPDIF status registers look like
+    #      0: 0000000011110000010010001010........$  with different volume levels set.
```

Comment:
I should switch to hex notation. Not sure why I wrote it like this.

Reply:
Fixed. Also changed the implementation of `parse_volume_from_status` a little bit.